sql/sql_parse.cc

```diff
@@ -6310,45 +6310,91 @@ case SQLCOM_PREPARE:
 
   if (! thd->in_sub_stmt)
   {
+#ifdef WITH_WSREP
+    int killed_errno;
+    bool is_error;
+    THD::killed_state killed;
+    bool has_sp_runtime_ctx;
+    wsrep_conflict_state conflict_state;
+
+    mysql_mutex_lock(&thd->LOCK_wsrep_thd);
+    conflict_state= wsrep_thd_conflict_state(thd);
+    has_sp_runtime_ctx= (thd->sp_runtime_ctx != NULL);
+    killed_errno= thd->killed_errno();
+    is_error= thd->is_error();
+    killed= thd->killed;
+    mysql_mutex_unlock(&thd->LOCK_wsrep_thd);
+
     /* report error issued during command execution */
-    if (thd->killed_errno())
+    if (killed_errno)
       thd->send_kill_message();
-    if (thd->is_error() || (thd->variables.option_bits & OPTION_MASTER_SQL_ERROR))
+    if (is_error || (thd->variables.option_bits & OPTION_MASTER_SQL_ERROR))
       trans_rollback_stmt(thd);
-#ifdef WITH_WSREP
-    else if (thd->sp_runtime_ctx &&
-             (thd->wsrep_conflict_state == MUST_ABORT ||
-              thd->wsrep_conflict_state == CERT_FAILURE ||
-              thd->wsrep_conflict_state == ABORTED))
+    else if (has_sp_runtime_ctx &&
+             (conflict_state == MUST_ABORT ||
+              conflict_state == CERT_FAILURE ||
+              conflict_state == ABORTED))
     {
+      trans_rollback_stmt(thd);
+
       /*
-        The error was cleared, but THD was aborted by wsrep and
-        wsrep_conflict_state is still set accordingly. This
-        situation is expected if we are running a stored procedure
-        that declares a handler that catches ER_LOCK_DEADLOCK error.
-        In which case the error may have been cleared in method
-        sp_rcontext::handle_sql_condition().
+        PXC-3243
+        If BF-aborted, do not clear the conflict-state/killed variables.
+        Let the error propagate upward (if this was called from an SP),
+        to ensure that the locks are cleared.
       */
-      trans_rollback_stmt(thd);
-      thd->wsrep_conflict_state= NO_CONFLICT;
-      thd->killed= THD::NOT_KILLED;
+      if (conflict_state != MUST_ABORT) {
```

Comment:
Let us keep it uniform with the rest of the code.
```suggestion
      if (conflict_state != MUST_ABORT)
      {
```

Reply:
+1

Reply:
fixed